logstash-core/src/main/java/org/logstash/common/BufferedTokenizerExt.java

```diff
@@ -77,22 +79,53 @@ public IRubyObject init(final ThreadContext context, IRubyObject[] args) {
     @SuppressWarnings("rawtypes")
     public RubyArray extract(final ThreadContext context, IRubyObject data) {
         final RubyArray entities = data.convertToString().split(delimiter, -1);
+        if (!bufferFullErrorNotified) {
+            input.clear();
+            input.addAll(entities);
+        } else {
+            // after a full buffer signal
+            if (input.isEmpty()) {
+                // after a buffer full error, the remaining part of the line, till next delimiter,
+                // has to be consumed, unless the input buffer doesn't still contain fragments of
+                // subsequent tokens.
+                entities.shift(context);
+                input.addAll(entities);
+            } else {
+                // merge last of the input with first of incoming data segment
+                if (!entities.isEmpty()) {
+                    RubyString last = ((RubyString) input.pop(context));
+                    RubyString nextFirst = ((RubyString) entities.shift(context));
+                    entities.unshift(last.concat(nextFirst));
+                    input.addAll(entities);
+                }
+            }
+        }
+
         if (hasSizeLimit) {
-            final int entitiesSize = ((RubyString) entities.first()).size();
+            if (bufferFullErrorNotified) {
+                bufferFullErrorNotified = false;
+                if (input.isEmpty()) {
+                    return RubyUtil.RUBY.newArray();
+                }
+            }
+            final int entitiesSize = ((RubyString) input.first()).size();
             if (inputSize + entitiesSize > sizeLimit) {
+                bufferFullErrorNotified = true;
+                headToken = new StringBuilder();
+                inputSize = 0;
+                input.shift(context); // consume the token fragment that generates the buffer full
                 throw new IllegalStateException("input buffer full");
             }
             this.inputSize = inputSize + entitiesSize;
         }
-        input.append(entities.shift(context));
-        if (entities.isEmpty()) {
+        headToken.append(input.shift(context)); // remove head
```

Comment:
```java

        if (input.getLength() < 2) {
            headToken.append(input.shift(context)); // remove head
            return RubyUtil.RUBY.newArray();
        } else {
            if (headToken.length() > 0) {
                headToken.append(input.shift(context)); // append buffer to first element and
                input.unshift(RubyUtil.toRubyObject(headToken.toString())); // reinsert it into the array
                headToken = new StringBuilder();
            }
            headToken.append(input.pop(context)); // put the leftovers in headToken for later
            inputSize = headToken.length();
            return input;
        }
```

Reply:
We can do that but complicates even more the flow. Before applying that maybe a measure of effective performance gain should be done.

Reply:
About question 1, I haven't done any performance check on this. However, we should define the use cases to check. The buffer full condition in the existing implementation at certain point didn't provide correct results, so if we use the original implementation as baseline, we aren't testing apples-to-apples. The only thing we can test is aggregation process, while any token size doesn't fail the size limit.
I could think to:
- tokens that are smaller then a full segment
- tokens that span multiple segments
- tokens that are always spread between the end of a preceding segment and the subsequent one.


On point number 2, this is a bug. The expectation for the client of this tokenizer is that lines (or tokens) bigger than the predefined limit generates an exception and that the normal flow is able to resume on the next `extract` call immediately after the token delimiter that marks the end of the offending token. So if the token that should trigger the error is not the first of a segment, is must still raise the error.

Reply:
@andsel given my proposal improves performance should we add it to this PR?

Reply:
yes, I missed to add it

Reply:
added with [99292ca](https://github.com/elastic/logstash/pull/16482/commits/99292ca04644d379a184a21c5f6c21a2ffe299ea)